tests/src/Kernel/UniqueEmailTest.php

```diff
@@ -0,0 +1,586 @@
+    $this->setUniqueEmailsConfig('none');
+
+    $existing_user = $this->createUser();
+
+    // Test validating a decoupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->getAccountName());
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Decoupled user fails validation.');
+
+    // Test validating a coupled user.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->getAccountName());
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Coupled user fails validation.');
+  }
+
+  /**
+   * Test the unique email validator in 'include' mode with no existing users.
+   *
+   * @covers ::validate
+   */
+  public function testModeIncludeNone() {
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    // Test validating a decoupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser();
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user passes validation.');
+
+    // Test validating a decoupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser();
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user passes validation.');
+
+    // Test validating a coupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(FALSE);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Coupled user passes validation.');
+
+    // Test validating a coupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(FALSE);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Coupled user passes validation.');
+  }
+
+  /**
+   * Test the unique email validator in 'include' mode with an existing decoupled
+   * user with no matching roles.
+   *
+   * @covers ::validate
+   */
+  public function testModeIncludeDecoupledWithoutRole() {
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    $existing_user = $this->createDecoupledUser();
+
+    // Test validating a decoupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user without role passes validation.');
+
+    // Test validating a decoupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user with role passes validation.');
```

Comment:
If the user has the role, it's email SHOULD be unique, so this should be assertNotEmpty()